 ---
 ---
   **Problem Formulation** 

 ---
 ---

*  **The input:** 
 *  Training dat (df) : it containd the features and the label data.
 *  Testing data (df2) : it contains the features.

* **the output**
 * it will be the label data (the raiting column) 

*  **Function used**  
 *   classification

 *   predection

*   **Challenges**

 *   The dataset is not clean, and we need some preprocessing depending on the models we choice,
 
 - changing the multipul  hyperparameter to give a good accuracy/f1 score

*    **the imapct** 
 *   the impact that the model will fits will on the training data and predect well on the test data or any data that we enter, And that will have a huge impact in real life if that uesd we will be able to predect right and proratize what is important.
 
 - we are going to be able to evaluate our mdels correctly and we can help the customer to find what they really want. 


*    **ideal solution**
 *   the ideal solutition happens when our model is able to classify and predict very well on any data the model gets , And also that happens when our data is cleaned perfectly  and 

 - it happens when choose the perfecrt model who can deals with type of dat and not to forget the gyperparameter tuning .


 







---
---
#**Importing libraries**

----

---

In [1]:
import pandas as pd 
import numpy as np 
from sklearn.model_selection import train_test_split
# import to RFE that return best columns
from sklearn.feature_selection import RFE
from sklearn.feature_selection import SelectKBest, mutual_info_classif

# import packages for hyperparameters tuning
from hyperopt import STATUS_OK, Trials, fmin, hp, tpe

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.model_selection import PredefinedSplit

In [2]:
#reading traingi data
data = pd.read_csv('/content/drive/MyDrive/DM_assignment2/train.csv')
#reading testing  data
test = pd.read_csv('/content/drive/MyDrive/DM_assignment2/test.csv')
id_ = test.id

---
---
#**over view of the data**

---
---

In [3]:
# to show the first 5 rows of data
data.head()

gender  idg  condtn  wave  round  position  positin1  order  partner  \
0       0    3       2    14     18         2       2.0     14       12   
1       1   14       1     3     10         2       NaN      8        8   
2       1   14       1    13     10         8       8.0     10       10   
3       1   38       2     9     20        18      13.0      6        7   
4       1   24       2    14     20         6       6.0     20       17   

     pid  ...  sinc3_3  intel3_3  fun3_3  amb3_3  attr5_3  sinc5_3  intel5_3  \
0  372.0  ...      NaN       NaN     NaN     NaN      NaN      NaN       NaN   
1   63.0  ...      8.0       8.0     7.0     8.0      NaN      NaN       NaN   
2  331.0  ...      NaN       NaN     NaN     NaN      NaN      NaN       NaN   
3  200.0  ...      9.0       8.0     8.0     6.0      NaN      NaN       NaN   
4  357.0  ...      NaN       NaN     NaN     NaN      NaN      NaN       NaN   

   fun5_3  amb5_3    id  
0     NaN     NaN  2583  
1     NaN     NaN  6830  
2     NaN     NaN  4840  
3     NaN     NaN  5508  
4     NaN     NaN  4828  

[5 rows x 192 columns]

In [4]:
# to show the first 5 rows of data
test.head()

gender  idg  condtn  wave  round  position  positin1  order  partner  \
0       0    5       2     2     16         3       NaN     13       13   
1       0   33       2    14     18         6       6.0      4        8   
2       1    6       2     9     20        10      16.0     15       19   
3       1   26       2     2     19        15       NaN      8       10   
4       0   29       2     7     16         7       7.0     10        5   

     pid  ...  sinc3_3  intel3_3  fun3_3  amb3_3  attr5_3  sinc5_3  intel5_3  \
0   52.0  ...      7.0       8.0     6.0     8.0      NaN      NaN       NaN   
1  368.0  ...      8.0       7.0     7.0     8.0      6.0      7.0       6.0   
2  212.0  ...      NaN       NaN     NaN     NaN      NaN      NaN       NaN   
3   30.0  ...      NaN       NaN     NaN     NaN      NaN      NaN       NaN   
4  162.0  ...      NaN       NaN     NaN     NaN      NaN      NaN       NaN   

   fun5_3  amb5_3    id  
0     NaN     NaN   934  
1     5.0     5.0  6539  
2     NaN     NaN  6757  
3     NaN     NaN  2275  
4     NaN     NaN  1052  

[5 rows x 191 columns]

In [5]:
# printing the shape of train and test data
print(data.shape)
print(test.shape)

(5909, 192)
(2469, 191)


In [6]:
#a full describetion of the data
data.describe()
# theis line to get the full decribtion even with string values 
#data.describe(include='all')

gender          idg       condtn         wave        round  \
count  5909.000000  5909.000000  5909.000000  5909.000000  5909.000000   
mean      0.505331    17.360298     1.824843    11.347436    16.850228   
std       0.500014    10.947542     0.380133     6.011495     4.389246   
min       0.000000     1.000000     1.000000     1.000000     5.000000   
25%       0.000000     8.000000     2.000000     7.000000    14.000000   
50%       1.000000    16.000000     2.000000    11.000000    18.000000   
75%       1.000000    26.000000     2.000000    15.000000    20.000000   
max       1.000000    44.000000     2.000000    21.000000    22.000000   

          position     positin1       order      partner          pid  ...  \
count  5909.000000  4591.000000  5909.00000  5909.000000  5901.000000  ...   
mean      9.001523     9.254846     8.91166     8.962938   283.733266  ...   
std       5.482368     5.611803     5.45710     5.500706   158.993002  ...   
min       1.000000     1.000000     1.00000     1.000000     1.000000  ...   
25%       4.000000     4.000000     4.00000     4.000000   153.000000  ...   
50%       8.000000     9.000000     8.00000     8.000000   280.000000  ...   
75%      13.000000    14.000000    13.00000    13.000000   409.000000  ...   
max      22.000000    22.000000    22.00000    22.000000   552.000000  ...   

           sinc3_3     intel3_3       fun3_3       amb3_3      attr5_3  \
count  2804.000000  2804.000000  2804.000000  2804.000000  1413.000000   
mean      8.105563     8.377318     7.644437     7.398716     6.799717   
std       1.601011     1.459013     1.757559     1.956924     1.535768   
min       2.000000     3.000000     2.000000     1.000000     2.000000   
25%       7.000000     8.000000     7.000000     6.000000     6.000000   
50%       8.000000     8.000000     8.000000     8.000000     7.000000   
75%       9.000000     9.000000     9.000000     9.000000     8.000000   
max      12.000000    12.000000    12.000000    12.000000    10.000000   

           sinc5_3     intel5_3       fun5_3       amb5_3           id  
count  1413.000000  1413.000000  1413.000000  1413.000000  5909.000000  
mean      7.631989     7.944798     7.162774     7.092711  4191.314943  
std       1.498024     1.320919     1.687431     1.713729  2408.009173  
min       2.000000     4.000000     1.000000     1.000000     0.000000  
25%       7.000000     7.000000     6.000000     6.000000  2124.000000  
50%       8.000000     8.000000     7.000000     7.000000  4210.000000  
75%       9.000000     9.000000     8.000000     8.000000  6266.000000  
max      10.000000    10.000000    10.000000    10.000000  8372.000000  

[8 rows x 184 columns]

In [7]:
#a full describetion of the data
test.describe()

gender          idg       condtn         wave        round  \
count  2469.000000  2469.000000  2469.000000  2469.000000  2469.000000   
mean      0.489267    17.247874     1.838396    11.359255    16.924261   
std       0.499986    10.926235     0.368162     5.959627     4.284307   
min       0.000000     1.000000     1.000000     1.000000     5.000000   
25%       0.000000     8.000000     2.000000     7.000000    14.000000   
50%       0.000000    16.000000     2.000000    11.000000    18.000000   
75%       1.000000    26.000000     2.000000    15.000000    20.000000   
max       1.000000    44.000000     2.000000    21.000000    22.000000   

          position     positin1        order      partner          pid  ...  \
count  2469.000000  1941.000000  2469.000000  2469.000000  2467.000000  ...   
mean      9.141353     9.392581     8.965978     8.965168   284.175922  ...   
std       5.592006     5.740292     5.525290     5.469045   157.636121  ...   
min       1.000000     1.000000     1.000000     1.000000     1.000000  ...   
25%       4.000000     4.000000     4.000000     4.000000   157.000000  ...   
50%       8.000000     9.000000     8.000000     8.000000   283.000000  ...   
75%      13.000000    14.000000    13.000000    13.000000   405.000000  ...   
max      22.000000    22.000000    22.000000    22.000000   552.000000  ...   

           sinc3_3     intel3_3       fun3_3       amb3_3     attr5_3  \
count  1170.000000  1170.000000  1170.000000  1170.000000  603.000000   
mean      8.064103     8.416239     7.693162     7.374359    6.834163   
std       1.632694     1.459546     1.713641     1.972877    1.439486   
min       2.000000     3.000000     2.000000     1.000000    2.000000   
25%       7.000000     8.000000     7.000000     6.000000    6.000000   
50%       8.000000     8.000000     8.000000     8.000000    7.000000   
75%       9.000000     9.000000     9.000000     9.000000    8.000000   
max      12.000000    12.000000    12.000000    12.000000   10.000000   

          sinc5_3    intel5_3      fun5_3      amb5_3           id  
count  603.000000  603.000000  603.000000  603.000000  2469.000000  
mean     7.575456    7.903814    7.137645    6.945274  4181.763062  
std      1.520249    1.387187    1.639208    1.724936  2444.455043  
min      2.000000    4.000000    1.000000    1.000000     3.000000  
25%      7.000000    7.000000    6.000000    6.000000  2040.000000  
50%      8.000000    8.000000    7.000000    7.000000  4137.000000  
75%      9.000000    9.000000    8.000000    8.000000  6316.000000  
max     10.000000   10.000000   10.000000   10.000000  8377.000000  

[8 rows x 183 columns]

In [8]:
#checking if there is any null values and getting the sum in every column
data.isnull().sum().sort_values(ascending=False)

num_in_3    5449
numdat_3    4849
expnum      4627
amb7_2      4519
sinc7_2     4519
            ... 
position       0
round          0
wave           0
condtn         0
id             0
Length: 192, dtype: int64

In [9]:
#getting a full sammry of the train data 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5909 entries, 0 to 5908
Columns: 192 entries, gender to id
dtypes: float64(173), int64(11), object(8)
memory usage: 8.7+ MB


we realized we have 8 objest values and we need to change their vlaues to numeric later.  

---
---

In [10]:
# getting a full summary of the test data 
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2469 entries, 0 to 2468
Columns: 191 entries, gender to id
dtypes: float64(173), int64(10), object(8)
memory usage: 3.6+ MB


the same as the training  data we found the same values.

----
---

In [11]:
# to get the sum of unique values in every column
data.nunique()

gender         2
idg           44
condtn         2
wave          21
round         15
            ... 
sinc5_3        9
intel5_3       7
fun5_3        10
amb5_3         9
id          5909
Length: 192, dtype: int64

In [12]:
#check for a clumn with only one unique values  
print(data.loc[:,data.nunique()==1].keys())

Index([], dtype='object')


In [13]:
# # #convert all object columns to categorical column
# data[data.select_dtypes(['object']).columns] = data.select_dtypes(['object']).apply(lambda x: x.astype('category'))

---
---
#**Preprocessing**

---

---
**After reading the data i realized:**

- There is a alot of missing values.
- String values
- colums that not really necessary like the ( id )  
---

**Here im planing to** :

1) split the training data to features and and label.

2) Droping the columns that doesn't affect in the features.

3) put the numeric values and categorical data speratly. and that will help working in the categorical data when we want to change it to numerical.

4) Full PIPLINE : IT has the main preprocessig feaatures such as: 
  - Fill Nan values 
  - standard scaler
  - one HotEncoding 
---
**the experimental protocol used is 'Hold out method'**

The following is the process of using the hold-out method for model evaluation:

- Split the dataset into two parts (preferably based on 70-30% split; However, the percentage split will vary)

- Train the model on the training dataset; While training the model, some fixed set of hyper parameters is selected.

- Test or evaluate the model on the held-out test dataset
Train the final model on the entire dataset to get a model which can generalize better on the unseen or future dataset. 
---

---
###**Splitting training data** 

----

In [14]:
#splitting the teaining data to x,y (features and label/output) 
#and also droppin the id fron the x
y = data['match'] # lower case for vector
X = data.drop(['id','match'], axis=1) # upper case for matrix
print('original shape', X.shape, y.shape)


original shape (5909, 190) (5909,)


---
###**Dropping the id coulum**

---

we are dropping the id column from the training data and testing data, it's not really necessry to use it so we can drop it 

 ---

In [15]:
# dropping id from testing data
test.drop('id',axis=1,inplace=True)
print('original shape', test.shape)

original shape (2469, 190)


In [16]:
# #convert all object columns to categorical column
# data[data.select_dtypes(['object']).columns] = data.select_dtypes(['object']).apply(lambda x: x.astype('category'))

 The data didv't need to be changed to categorical cuz it gave the same results as it's onbject so there is no need to change it.

 ---- 


###**extract numeric and categorical feature**

---

---
In this step we are splitting the numeric and the categorical values to be able to deal with them later when we want to change the catigorical to numeric.

---

In [17]:
# we extract numeric features and categorical features names
# for later use

# numeric features can be selected by: (based on the data.info() output )
features_numeric = list(X.select_dtypes(include=['float64', 'int64']))

# categorical features can be selected by: (based on the df2.info() output )
features_categorical = list(X.select_dtypes(include=['object']))

print('numeric features:', features_numeric)
print('categorical features:', features_categorical)

numeric features: ['gender', 'idg', 'condtn', 'wave', 'round', 'position', 'positin1', 'order', 'partner', 'pid', 'int_corr', 'samerace', 'age_o', 'race_o', 'pf_o_att', 'pf_o_sin', 'pf_o_int', 'pf_o_fun', 'pf_o_amb', 'pf_o_sha', 'attr_o', 'sinc_o', 'intel_o', 'fun_o', 'amb_o', 'shar_o', 'like_o', 'prob_o', 'met_o', 'age', 'field_cd', 'race', 'imprace', 'imprelig', 'goal', 'date', 'go_out', 'career_c', 'sports', 'tvsports', 'exercise', 'dining', 'museums', 'art', 'hiking', 'gaming', 'clubbing', 'reading', 'tv', 'theater', 'movies', 'concerts', 'music', 'shopping', 'yoga', 'exphappy', 'expnum', 'attr1_1', 'sinc1_1', 'intel1_1', 'fun1_1', 'amb1_1', 'shar1_1', 'attr4_1', 'sinc4_1', 'intel4_1', 'fun4_1', 'amb4_1', 'shar4_1', 'attr2_1', 'sinc2_1', 'intel2_1', 'fun2_1', 'amb2_1', 'shar2_1', 'attr3_1', 'sinc3_1', 'fun3_1', 'intel3_1', 'amb3_1', 'attr5_1', 'sinc5_1', 'intel5_1', 'fun5_1', 'amb5_1', 'attr', 'sinc', 'intel', 'fun', 'amb', 'shar', 'like', 'prob', 'met', 'match_es', 'attr1_s', 'sin

---
###**FuLL PIPILINE**

---

This is the most important step in the preprocessing, it's where i can Fill the NAN values and also convert the categorical values into numerical by using onehotEncoding, and also we scale the data by StandardScaler.

 ---

- The first pipe line for numeric feature preprocessing. 
   - **SimpleImputer** : we use it to fill the Nan values.
   - **StandardScaler** : to scale the numeric values.
- The second pip line for object feature preprocessing.
  - **SimpleImputer** : we use it to fill the Nan values.
  - **OneHotEncoder** : we use it to convert the objest to numeric values.

- The last column in the full pipline we use it to call the first and second pip line.


In [18]:
#import all the libraries that we use it the follwing steps 
import numpy as np
from sklearn.compose import ColumnTransformer
from sklearn.datasets import fetch_openml
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from xgboost.sklearn import XGBClassifier

np.random.seed(0)

# define a pipe line for numeric feature preprocessing
transformer_numeric = Pipeline(
    steps=[
        ('imputer', SimpleImputer()),
        ('scaler', StandardScaler())]
)

# define a pipe line for object feature preprocessing
transformer_categorical = Pipeline(
    steps=[
        ('imputer', SimpleImputer(strategy='constant')),
        ('onehot', OneHotEncoder(handle_unknown='ignore'))
    ]
)

# define the preprocessor 
preprocessor = ColumnTransformer(
    transformers=[
        ('num', transformer_numeric, features_numeric),
        ('cat', transformer_categorical, features_categorical)
    ]
)


---
---
#**modeling** 

---

---

--- 
**my plans for thr trials**

im planning to use:

- Random forst with Grid search  
- Logistic regresstion with grid earch
- losgistic regresstion with validation set
- losgistic regresstion with random forest
- losgistic regresstion with Bayes search
- logistic regression with feature selection 
- XGBClassifier with grid earch
- XGBClassifier with validation set
- XGBClassifier with  random forest
- XGBClassifier regresstion with Bayes search



##**Trial_1**

---


###Random forest with Grid Search

---
Random forest is a flexible, easy to use machine learning algorithm that produces, even without hyper-parameter tuning
 
im expecting a great result most of the time. It is also one of the most used algorithms, because of its simplicity and diversity.

And also im using Hyperparameter tuning : **Grid search** and it allowed us to narrow down the range for each hyperparameter.

---

**Hyperparameters used in Grid search** :

---
---
- **n_estimators** : A object of that type is instantiated for each grid point. This is assumed to implement the scikit-learn estimator interface. Either estimator needs to provide a score function, or scoring must be passed.

- **max_depth** :the depth of the tree.

- **n_jobs** : Number of jobs to run in paralle.

- **cv** : int, cross-validation generator or an iterable

- **verbose** : Controls the verbosity: the higher, the more messages.




In [19]:
# combine the preprocessor with the model as a full tunable pipeline
# here we call the full pip line with the classifier RandomForestClassifier
full_pipline = Pipeline(steps=[('preprocessor', preprocessor),('my_classifier', RandomForestClassifier(),)])


---
We fit the model with full_pipline so all the steps excist in the pipline will be performed on the x,y (the features and the label)

---

In [20]:
# The pipeline object can be used like any sk-learn model
full_pipline = full_pipline.fit(X, y)

---
This step is really important bescause in here we perform the same preprocesing

 steps in the test file and that's by predictiong using the **full_pipline**


In [21]:
# predect the output for the test data using the full_pipline do all the preprocessing will be done on the test file too
y_test_pred = full_pipline.predict_proba(test)

--- 
---

**Hyperparameters used in Grid search** :

---
---
- **n_estimators** : A object of that type is instantiated for each grid point. This is assumed to implement the scikit-learn estimator interface. Either estimator needs to provide a score function, or scoring must be passed.

- **max_depth** :the depth of the tree.

- **n_jobs** : Number of jobs to run in paralle.

- **cv** : int, cross-validation generator or an iterable

- **verbose** : Controls the verbosity: the higher, the more messages.




In [22]:

# here we specify the search space
# `__` denotes an attribute of the preceeding name
# (e.g. my_classifier__n_estimators means the `n_estimators` param for `my_classifier`)
param_grid = {
    'preprocessor__num__imputer__strategy': ['mean'],
    # my_classifier__n_estimators points to my_classifier->n_estimators y
    'my_classifier__n_estimators': [20, 30, 40],  
    'my_classifier__max_depth':[10, 20, 30]       
}

# cv=2 means two-fold cross-validation
# n_jobs means the cucurrent number of jobs
# (on colab since we only have two cpu cores, we set it to 2)
grid_search = GridSearchCV(
    full_pipline, param_grid, cv=2, verbose=1, n_jobs=2, 
    scoring='roc_auc')

grid_search.fit(X, y)

print('best score {}'.format(grid_search.best_score_))
print('best score {}'.format(grid_search.best_params_))

Fitting 2 folds for each of 9 candidates, totalling 18 fits
best score 0.831188762194597
best score {'my_classifier__max_depth': 10, 'my_classifier__n_estimators': 40, 'preprocessor__num__imputer__strategy': 'mean'}


- After performaing the model with the Grid Search, i realized that the model perform good.it gave me an accuracy = 0.827.

- **"my thoughts"** 

 - i thought the modell will give me a higer accuracy than that.

---

In [23]:
# predicting the output for the the test value 
y_test_pred_ = grid_search.predict_proba(test)

In [24]:
# for writting the results in a file on drive
dummy = pd.DataFrame({'id':id_,'match': y_test_pred[:,1]})
dummy.to_csv('/content/drive/MyDrive/DM_assignment2/assignment2_RF.csv', encoding='utf-8', index=False)


#**With logistic Regression**


##**Trial_2**

---

###Grid Search with Cross-validation

---
Logistic regression is not a classifier. It is a probability/risk estimator, it allows for and expects "close calls". It will lead to optimum decision making because it does not try to trick the predictive signal into incorporating a utility function that is implicit whenever you classify observations.

---

im expecting cross validation with the logistic regression to give a high a ccuracy than the grid search and validation set since it does not try to trick the predictive signal into incorporating.

---

**Hyperparameters used in Grid search** :

---
- **penalty** : the regulrization to control overfitting 

- **C** : is inverse of regularization, the larger the C, the smaller is regularization, means that your algo is more prone to overfit the data

- **n_jobs** : Number of jobs to run in paralle.

- **cv** : int, cross-validation generator or an iterable

- **verbose** : Controls the verbosity: the higher, the more messages.


In [25]:
# combine the preprocessor with the model as a full tunable pipeline
# here we call the full pip line with the classifier RandomForestClassifier
full_pipline = Pipeline(steps=[('preprocessor', preprocessor),('my_classifier', LogisticRegression(),)])

---
We fit the model with full_pipline so all the steps excist in the pipline will be performed on the x,y (the features and the label)

---

In [ ]:
# The pipeline object can be used like any sk-learn model
full_pipline = full_pipline.fit(X, y)

--- 
In order to find the optimal value of the tuning parameter, a grid search needs to be conducted, and the optimal model is the one with the lowest GBIC (Generalized Bayesian Information Criterion)

---

**Hyperparameters used in Grid search** :

---
---
- **penalty** : the regulrization to control overfitting 

- **C** : is inverse of regularization, the larger the C, the smaller is regularization, means that your algo is more prone to overfit the data

- **n_jobs** : Number of jobs to run in paralle.

- **cv** : int, cross-validation generator or an iterable

- **verbose** : Controls the verbosity: the higher, the more messages.




In [27]:
param_grid = {
    'preprocessor__num__imputer__strategy': ['mean'],
    'my_classifier__penalty' : ['l2'],  
    'my_classifier__C' : np.logspace(-4, 4, 20),
    'my_classifier__solver': ['newton-cg']      
}

grid_search = GridSearchCV(
    full_pipline, param_grid, cv=2, verbose=1, n_jobs=2, 
    scoring='roc_auc')

grid_search.fit(X, y)

print('best score {}'.format(grid_search.best_score_))
print('best score {}'.format(grid_search.best_params_))

Fitting 2 folds for each of 20 candidates, totalling 40 fits
best score 0.8568706602967452
best score {'my_classifier__C': 0.004832930238571752, 'my_classifier__penalty': 'l2', 'my_classifier__solver': 'newton-cg', 'preprocessor__num__imputer__strategy': 'mean'}


- After performaing the model with the Grid Search, i realized that the model perform IS good.it gave me an accuracy = 0.856.

- **"my thoughts"** 

 - i thought the modell will give me a higer accuracy than 0.856.

---

In [28]:
# predict the the test data
y_test_pred_ = grid_search.predict_proba(test)

In [29]:
dummy = pd.DataFrame({'id':id_,'match': y_test_pred_[:,1]})
dummy.to_csv('/content/drive/MyDrive/DM_assignment2/assignment2_LG.csv', encoding='utf-8', index=False)


##**Trial_3**

---

### Grid Search with validation set

---
in validation set we are splitting data to have a validation set 

---
im expecting this trial will not going to give a pretty got accuracy comparing to cross validation.

---

In [30]:
# Further split the original training set to a train and a validation set
X_train2, X_val, y_train2, y_val = train_test_split(
    X, y, train_size = 0.8, stratify = y, random_state = 2022)

# Create a list where train data indices are -1 and validation data indices are 0
# X_train2 (new training set), X_train
split_index = [-1 if x in X_train2.index else 0 for x in X.index]

# Use the list to create PredefinedSplit
pds = PredefinedSplit(test_fold = split_index)

grid_search = GridSearchCV(
    full_pipline, param_grid, cv=pds, verbose=1, n_jobs=2, 
    scoring='roc_auc')

# here we still use X_train; but the grid search model
# will use our predefined split internally to determine 
# which sample belongs to the validation set
grid_search.fit(X, y)

print('best score {}'.format(grid_search.best_score_))
print('best score {}'.format(grid_search.best_params_))

Fitting 1 folds for each of 20 candidates, totalling 20 fits
best score 0.860885275519422
best score {'my_classifier__C': 0.012742749857031334, 'my_classifier__penalty': 'l2', 'my_classifier__solver': 'newton-cg', 'preprocessor__num__imputer__strategy': 'mean'}


- After performaing the model with the validation set, i realized that the model perform pretty well.it gave me an accuracy = 0.860. and that wasn't expecting.

- **"my thoughts"** 

 - i thought the modell will give me a lower accuracy but it gave a pretty good one, accuracy = 0.860.

---

In [31]:
# predict the test data
y_test_pred_ = grid_search.predict_proba(test)

In [32]:
dummy = pd.DataFrame({'id':id_,'match': y_test_pred_[:,1]})
dummy.to_csv('/content/drive/MyDrive/DM_assignment2/assignment2_LG_gsv.csv', encoding='utf-8', index=False)


##**Trial_4**

---

### Random Search

---
Random search is a basic improvement to grid search. The search process continues until the predetermined iteration number or the desired accuracy are reached. Random search is similar to grid search, but has been proven to produce better results

- it' expected that it will do preety well

---

In [33]:
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.model_selection import PredefinedSplit

param_grid = {
    'preprocessor__num__imputer__strategy': ['mean'],
    # preprocessor__num__imputer__strategy points to preprocessor->num (a Pipeline)-> imputer -> strategy
    'my_classifier__penalty' : ['l2'],  
    'my_classifier__C' : np.logspace(-4, 4, 20),
    'my_classifier__solver': ['newton-cg']      
} 

grid_search = RandomizedSearchCV(
    full_pipline, param_grid, cv=pds, verbose=1, n_jobs=2, 
    # number of random trials
    n_iter=10,
    scoring='roc_auc')

grid_search.fit(X, y)

print('best score {}'.format(grid_search.best_score_))
print('best score {}'.format(grid_search.best_params_))

Fitting 1 folds for each of 10 candidates, totalling 10 fits
best score 0.8604079412006241
best score {'preprocessor__num__imputer__strategy': 'mean', 'my_classifier__solver': 'newton-cg', 'my_classifier__penalty': 'l2', 'my_classifier__C': 0.03359818286283781}


---
- After performaing the model with the validation set, i realized that the model perform pretty well.it gave me an accuracy = 0.860. and that wasn't expecting.

- **"my thoughts"** 

 - i thought the modell will give me a lower accuracy but it gave a pretty good one, accuracy = 0.860.

---

In [34]:
# predict the test data
y_test_pred_ = grid_search.predict_proba(test)

In [35]:
dummy = pd.DataFrame({'id':id_,'match': y_test_pred_[:,1]})
dummy.to_csv('/content/drive/MyDrive/DM_assignment2/assignment2_LG_gs.csv', encoding='utf-8', index=False)


##**Trial_5**

---

### Bayes Search

---

Bayesian Optimization, By assuming an unknown objective function that gets an input, the initial goal was to discover an optimal solution that maximises or decreases the function value. Bayesian optimization is an iterative process

- it's expected that it will do the best performance 

---

In [36]:
!pip install scikit-optimize

In [37]:
# Let's try this with SVM model
from skopt import BayesSearchCV

param_grid = {
    'preprocessor__num__imputer__strategy': ['mean'],
    # preprocessor__num__imputer__strategy points to preprocessor->num (a Pipeline)-> imputer -> strategy
    'my_classifier__penalty' : [ 'l2'],  
     # my_classifier__n_estimators points to my_classifier->n_estimators 
    'my_classifier__C' : np.logspace(-4, 4, 20),
}
bayes_search = BayesSearchCV(full_pipline, param_grid, cv=pds, verbose=1, n_jobs=2, 
    # number of random trials
    n_iter=10,
    scoring='roc_auc')


bayes_search.fit(X, y)

print('best score {}'.format(bayes_search.best_score_))
print('best score {}'.format(bayes_search.best_params_))


Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
best score 0.8608955407735895
best score OrderedDict([('my_classifier__C', 0.012742749857031334), ('my_classifier__penalty', 'l2'), ('preprocessor__num__imputer__strategy', 'mean')])


---
- After performaing the model with the validation set, i realized that the model perform pretty well.it gave me an accuracy = 0.860. and that wasn't expecting.

- **"my thoughts"** 

 - i thought the modell will give me a lower accuracy but it gave a pretty good one, accuracy = 0.860.

---

In [38]:
# predict the test data
y_test_pred_ = bayes_search.predict_proba(test)

In [39]:
dummy = pd.DataFrame({'id':id_,'match': y_test_pred_[:,1]})
dummy.to_csv('/content/drive/MyDrive/DM_assignment2/assignment2_LG_bs.csv', encoding='utf-8', index=False)


##**Trial_6**

---

### feature selection

---
- it's expecting that this trial will give the highest performance and a high accuracy

---

In [40]:

# combine the preprocessor with the model as a full tunable pipeline
# we gave them a name so we can set their hyperparameters
full_pipline = Pipeline( steps=[('preprocessor', preprocessor),('selector', SelectKBest(mutual_info_classif, k=5)),('my_classifier', LogisticRegression(),)])


In [ ]:
# The pipeline object can be used like any sk-learn model
full_pipline = full_pipline.fit(X, y)

In [42]:
# predect the output for the test data using the full_pipline do all the preprocessing will be done on the test file too
y_test_pred = full_pipline.predict_proba(test)

In [43]:

# here we specify the search space
# `__` denotes an attribute of the preceeding name
# (e.g. my_classifier__n_estimators means the `n_estimators` param for `my_classifier`)
param_grid = {
    'preprocessor__num__imputer__strategy': ['mean'],
    'selector__k': [100, 150, 160],
    'my_classifier__penalty' : ['l2'],   
    'my_classifier__C' : np.logspace(-4, 4 ),
    'my_classifier__solver': ['newton-cg']       
}

# cv=2 means two-fold cross-validation
# n_jobs means the cucurrent number of jobs
# (on colab since we only have two cpu cores, we set it to 2)
grid_search = GridSearchCV(
    full_pipline, param_grid, cv=2, verbose=1, n_jobs=2, 
    scoring='roc_auc')

grid_search.fit(X, y)

print('best score {}'.format(grid_search.best_score_))
print('best score {}'.format(grid_search.best_params_))

Fitting 2 folds for each of 150 candidates, totalling 300 fits


/usr/local/lib/python3.7/dist-packages/sklearn/metrics/cluster/_supervised.py:65: UserWarning: Clustering metrics expects discrete values but received continuous values for label, and binary values for target
  warnings.warn(msg, UserWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/cluster/_supervised.py:65: UserWarning: Clustering metrics expects discrete values but received continuous values for label, and binary values for target
  warnings.warn(msg, UserWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/cluster/_supervised.py:65: UserWarning: Clustering metrics expects discrete values but received continuous values for label, and binary values for target
  warnings.warn(msg, UserWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/cluster/_supervised.py:65: UserWarning: Clustering metrics expects discrete values but received continuous values for label, and binary values for target
  warnings.warn(msg, UserWarning)
/usr/local/lib/python3.7/dis

best score 0.8560072217605473
best score {'my_classifier__C': 0.0062505519252739694, 'my_classifier__penalty': 'l2', 'my_classifier__solver': 'newton-cg', 'preprocessor__num__imputer__strategy': 'mean', 'selector__k': 100}


---
- After performaing the model with the validation set, i realized that the model perform quite bad .it gave me an accuracy = 0.850. and that wasn't expecting.

- **"my thoughts"** 

 - i thought the modell will give me a hugher accuracy
---

In [44]:
y_test_pred_ = grid_search.predict_proba(test)

**Observation**:  


After using th feature selection we realize that that didn't affect the model acuracy . so using the feature selection with the LOGISTIC regression model didn'y really go as expected.the accuracy of the model decreased from 0.86664 to 0.86365

#**With XGBClassifier**

---
##**Trial_7** 

---

 in this trial and the upcomming trials im using the **XGBClassifier**

--- 

**Hyperparameters used in Grid search** :

---
- **max_depth** : Maximum depth of a tree. Increasing this value will make the model more complex and more likely to overfit 

- **subsamble** : Same as the subsample of GBM. Denotes the fraction of observations to be randomly samples for each tree.

- **nthread** : the number og threads to use for loading data

- **colsample_bytree** : corresponds to the fraction of features (the columns) to use. By default it is set to 1 meaning that we will use all features.

- **n_estimators** : A object of that type is instantiated for each grid point. This is assumed to implement the scikit-learn estimator interface. Either estimator needs to provide a score function, or scoring must be passed.

- **n_jobs** : Number of jobs to run in paralle.

- **cv** : int, cross-validation generator or an iterable

- **verbose** : Controls the verbosity: the higher, the more messages.

###**Grid Search with Cross-validation**

In [45]:
# calling the full pipline
full_pipline = Pipeline(steps=[('preprocessor', preprocessor),('my_classifier', XGBClassifier(),)])

In [46]:
# The pipeline object can be used like any sk-learn model
full_pipline = full_pipline.fit(X, y)

In [47]:
# predect the output for the test data using the full_pipline do all the preprocessing will be done on the test file too
y_test_pred = full_pipline.predict_proba(test)

--- 

**Hyperparameters used in Grid search** :

---
- **max_depth** : Maximum depth of a tree. Increasing this value will make the model more complex and more likely to overfit 

- **subsamble** : Same as the subsample of GBM. Denotes the fraction of observations to be randomly samples for each tree.

- **nthread** : the number og threads to use for loading data

- **colsample_bytree** : corresponds to the fraction of features (the columns) to use. By default it is set to 1 meaning that we will use all features.

- **n_estimators** : A object of that type is instantiated for each grid point. This is assumed to implement the scikit-learn estimator interface. Either estimator needs to provide a score function, or scoring must be passed.

- **n_jobs** : Number of jobs to run in paralle.

- **cv** : int, cross-validation generator or an iterable

- **verbose** : Controls the verbosity: the higher, the more messages.

In [48]:

# here we specify the search space
# `__` denotes an attribute of the preceeding name
# (e.g. my_classifier__n_estimators means the `n_estimators` param for `my_classifier`)
param_grid = {
        "my_classifier__max_depth" : [6],
        'my_classifier__subsamble' : [0.9],
        'my_classifier__nthread' :[5],
        'my_classifier__colsample_bytree' :[0.9],
        'my_classifier__learning_rate' :[0.005],
        'my_classifier__n_estimators':  [2000]
}

# cv=2 means two-fold cross-validation
# n_jobs means the cucurrent number of jobs
# (on colab since we only have two cpu cores, we set it to 2)
grid_search = GridSearchCV(
    full_pipline, param_grid, cv=5, verbose=1, n_jobs=2, 
    scoring='roc_auc')

grid_search.fit(X, y)

print('best score {}'.format(grid_search.best_score_))
print('best score {}'.format(grid_search.best_params_))

Fitting 5 folds for each of 1 candidates, totalling 5 fits
best score 0.8843569370157903
best score {'my_classifier__colsample_bytree': 0.9, 'my_classifier__learning_rate': 0.005, 'my_classifier__max_depth': 6, 'my_classifier__n_estimators': 2000, 'my_classifier__nthread': 5, 'my_classifier__subsamble': 0.9}


In [49]:
# grid_search.fit(X,y)

---
- After performaing the model i realized that the model perform pretty good.it gave me a high accuracy = 0.884.

- **"my thoughts"** 

 - the classifer handel the data pretty well and it was able to perform well with the hyperparameter tuning 

---

In [50]:
#predict the test dat 
y_test_pred_ = grid_search.predict_proba(test)

In [51]:
dummy = pd.DataFrame({'id':id_,'match': y_test_pred_[:,1]})
dummy.to_csv('/content/drive/MyDrive/DM_assignment2/assignment2_XGB2.csv', encoding='utf-8', index=False)

---
##**Trial_8**

---

---
### Grid Search with validation set
---

---
in validation set we are splitting data to have a validation set 

---
im expecting this trial will not going to give a pretty got accuracy comparing to cross validation.

---

In [52]:
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.model_selection import PredefinedSplit

# Further split the original training set to a train and a validation set
X_train1, X_val, y_train1, y_val = train_test_split(
    X, y, train_size = 0.8, stratify = y, random_state = 2022)

# Create a list where train data indices are -1 and validation data indices are 0
# X_train2 (new training set), X_train
split_index = [-1 if x in X_train1.index else 0 for x in X.index]

# Use the list to create PredefinedSplit
pds = PredefinedSplit(test_fold = split_index)

grid_search = GridSearchCV(
    full_pipline, param_grid, cv=pds, verbose=1, n_jobs=2, 
    scoring='roc_auc')

# here we still use X_train; but the grid search model
# will use our predefined split internally to determine 
# which sample belongs to the validation set
grid_search.fit(X, y)

print('best score {}'.format(grid_search.best_score_))
print('best score {}'.format(grid_search.best_params_))

Fitting 1 folds for each of 1 candidates, totalling 1 fits
best score 0.8797990063233966
best score {'my_classifier__colsample_bytree': 0.9, 'my_classifier__learning_rate': 0.005, 'my_classifier__max_depth': 6, 'my_classifier__n_estimators': 2000, 'my_classifier__nthread': 5, 'my_classifier__subsamble': 0.9}


---
- After performaing the model i realized that the model perform pretty good.it gave me a good accuracy = 0.879.

---
##**Trial_9**

---

---
###Random Search
---

---
Random search is a basic improvement to grid search. The search process continues until the predetermined iteration number or the desired accuracy are reached. Random search is similar to grid search, but has been proven to produce better results

- it' expected that it will do preety well

---

In [53]:
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.model_selection import PredefinedSplit


grid_search = RandomizedSearchCV(
    full_pipline, param_grid, cv=pds, verbose=1, n_jobs=2, 
    # number of random trials
    n_iter=30,
    scoring='roc_auc')

grid_search.fit(X, y)

print('best score {}'.format(grid_search.best_score_))
print('best score {}'.format(grid_search.best_params_))
print('>>>>>>>>>> pds = ', pds)

Fitting 1 folds for each of 1 candidates, totalling 1 fits


/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_search.py:296: UserWarning: The total space of parameters 1 is smaller than n_iter=30. Running 1 iterations. For exhaustive searches, use GridSearchCV.
  UserWarning,


best score 0.8797990063233966
best score {'my_classifier__subsamble': 0.9, 'my_classifier__nthread': 5, 'my_classifier__n_estimators': 2000, 'my_classifier__max_depth': 6, 'my_classifier__learning_rate': 0.005, 'my_classifier__colsample_bytree': 0.9}
>>>>>>>>>> pds =  PredefinedSplit(test_fold=array([-1, -1, ..., -1, -1]))


---
- After performaing the model i realized that the model perform pretty good.it gave me a good accuracy = 0.879.

---



In [54]:
# predection on the test data
y_test_pred_ = grid_search.predict_proba(test)

In [55]:
dummy = pd.DataFrame({'id':id_,'match': y_test_pred_[:,1]})
dummy.to_csv('/content/drive/MyDrive/DM_assignment2/assignment2_XGB_R.csv', encoding='utf-8', index=False)

---
##**Trial_10**

---

---
### Bayesian Search
---

---

Bayesian Optimization, By assuming an unknown objective function that gets an input, the initial goal was to discover an optimal solution that maximises or decreases the function value. Bayesian optimization is an iterative process

- it's expected that it will do the best performance 

---

In [56]:
# Let's try this with SVM model
from skopt import BayesSearchCV

bayes_search = BayesSearchCV(full_pipline, param_grid, cv=pds, verbose=1, n_jobs=2, 
    # number of random trials
    n_iter=10,
    scoring='roc_auc')


bayes_search.fit(X, y)

print('best score {}'.format(bayes_search.best_score_))
print('best score {}'.format(bayes_search.best_params_))


Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
Fitting 1 folds for each of 1 candidates, totalling 1 fits
best score 0.8797990063233966
best score OrderedDict([('my_classifier__colsample_bytree', 0.9), ('my_classifier__learning_rate', 0.005), ('my_classifier__max_depth', 6), ('my_classifier__n_estimators', 2000), ('my_classifier__nthread', 5), ('my_classifier__subsamble', 0.9)])


---
- After performaing the model i realized that the model perform pretty good.it gave me a good accuracy = 0.8790.

---



In [57]:
# predict on the test data
y_test_pred_ = bayes_search.predict_proba(test)

In [58]:
dummy = pd.DataFrame({'id':id_,'match': y_test_pred_[:,1]})
dummy.to_csv('/content/drive/MyDrive/DM_assignment2/assignment2_XGB_BS.csv', encoding='utf-8', index=False)

#**Questions**

--- 
**1- Why a simple linear regression model (without any activation function) is not good for classification task, compared to Perceptron/Logistic regression?**

---

There are two things that explain why Linear Regression is not suitable for classification. The first one is that Linear Regression deals with continuous values whereas classification problems mandate discrete values.
The second problem is regarding the shift in threshold value when new data points are added. Let us take a simple Linear Regression example and fit a line to it. The below graph shows the best fit line. To make the explanation a bit more simple let us take an example of a classification problem in healthcare. Basically, our aim here is to classify whether a person is sick or not.
Technically the hypothesis function for linear regression can be used for logistic regression also but there is a problem.

---
**2-What's a decision tree and how it is different to a logistic regression model?**

---


- The decision trees is a quite popular data mining technique (Hastie, Tibshirani and Friedman, 2009), which is easey of use, robustness with missing data and ease of interpretability. Generally, decision trees are flexible.

- Regression models are relatively inflexible, for example, you have to add additional terms, i.e. interaction terms,polynomial terms. 
And decision trees can deal with missing values without imputation, while regression modelusually has to impute missing values before building the model, although there is not missing data in our data set.
And decision trees are nonparametric and highly robust, while regression models are parametric and sensitive to influencing points. 
We use the Tree node in Enterprise Miner to analyze the transportation data. The Gini reduction method is chosen as the splitting criterion.



---
**3-What's the difference between grid search and random search?**

---

- The grid search is an automatic search that searches an area within a range in a grid format.

- Grid search is the primary method for hyperparameter optimization and thoroughly searches hyperparameter sets in a custom range.

- Grid search can be applied to multiple hyper parameters. Grid search is the simplest search algorithm leading to the most accurate predictions and users can
always find the best combination.


- A random search algorithm has been proposed to improve the problem in grid search. Random search tries any combination of ranges of values.
Compare with grid search, random search can be considered more efficient in high-dimensional space. But random search is unreliable for training
some complex models. Bayesian method is recently drawing attention to solve this kind of problem.
It combines prior information about the unknown function with sample information, to obtain posterior information of the function distribution by using Bayesian formula.

- Random search is a basic improvement to grid search. The search process continues until the predetermined iteration number or the desired accuracy are reached. Random search is similar to grid search, but has been proven to produce
better results due to the following two benefits:

 - **First**, random search can perform better, especially if some hyper-parameters are not evenly distributed. In this search pattern, random searches are relatively more likely to find the optimal configuration than grid searches. 
     
  -  **Secondly**, it is not a promise that random searches will give you the best value, but you can get good value. And the more time
you spend, the more likely you are to find a better set of hyper-parameters. Conversely, for grid search, we cannot guarantee better results even with long search times. In most cases, random search is more effective than grid search, but it is still a computationally intensive method. Random searches generally require more time and computational resources than other search methods.

---
**4-What's the difference between bayesian search and random search?**

---

- Random search is a basic improvement to grid search. The search process continues until the predetermined iteration number or the desired accuracy are reached. Random search is similar to grid search, but has been proven to produce
better results due to the following two benefits:

 - **First**, random search can perform better, especially if some hyper-parameters are not evenly distributed. In this search pattern, random searches are relatively more likely to find the optimal configuration than grid searches. 
     
  -  **Secondly**, it is not a promise that random searches will give you the best value, but you can get good value. And the more time
you spend, the more likely you are to find a better set of hyper-parameters. Conversely, for grid search, we cannot guarantee better results even with long search times. In most cases, random search is more effective than grid search, but it is still a computationally intensive method. Random searches generally require more time and computational resources than other search methods.

- Bayesian Optimization, By assuming an unknown objective function that gets an input, the initial goal was to discover an optimal solution that maximises or decreases the function value. Bayesian optimization is an iterative process.

---
**refreances**

----
- https://www.mwsug.org/proceedings/2009/stats/MWSUG-2009-D02.pdf

- https://www.koreascience.or.kr/article/JAKO202009252091916.pdf